File: dss-service/backend/app/api/v1/endpoints/dss.py
"""DSS endpoints under /api/v1/dss."""
import os
from typing import List

import numpy as np
from fastapi import APIRouter, HTTPException

from schemas.dss_schema import (
    AHPResult,
    DSSRequest,
    WSSResult,
)
from services.dss_service import MotorbikeDSSService

router = APIRouter(prefix="/dss", tags=["DSS"])

# Instantiate the service once. Defaults resolve data files under backend/data/
# relative to services/dss_service.py.
service = MotorbikeDSSService()


@router.get("/default-criteria", summary="Default criteria (weights + benefit/cost)")
def default_criteria():
    """Return the 7 default criteria synced from criteria_weights.csv.

    Each entry exposes the column key, Vietnamese name, the default normalized
    weight (from criteria_weights.csv) and whether it is Benefit (maximize)
    or Cost (minimize).
    """
    return {
        "criteria": service.default_criteria(),
        "n_criteria": len(service.criteria_cols),
        "weights": service.default_weights,
        "benefit_mask": service.benefit_mask,
        "source": os.path.basename(service.criteria_path),
    }


@router.get("/criteria", summary="List evaluation criteria")
def list_criteria():
    return {"criteria": service.criteria, "n": len(service.criteria)}


@router.get("/brands", summary="List brands")
def list_brands():
    return {"brands": sorted(service.df_raw["brand"].dropna().unique().tolist())}


@router.post("/ahp", response_model=AHPResult, summary="Run AHP")
def run_ahp(weights_matrix: List[List[float]]):
    try:
        matrix = np.array(weights_matrix, dtype=float)
    except (TypeError, ValueError):
        raise HTTPException(status_code=400, detail="Invalid matrix values.")
    try:
        return service.weights_from_ahp(matrix)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/run", response_model=WSSResult, summary="Run DSS (AHP+TOPSIS)")
def run_dss(req: DSSRequest):
    pairwise = None
    if req.pairwise_matrix:
        pairwise = np.array(req.pairwise_matrix, dtype=float)

    if req.weights is None and pairwise is None:
        # Fall back to default criteria weights.
        pass

    try:
        result = service.run_dss(
            user_weights=req.weights,
            criteria_pairwise=pairwise,
            max_price=req.max_price_vnd,
            powertrain=req.powertrain,
            brand_list=req.brand_list,
            vehicle_type=req.vehicle_type,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    if result["status"] == "empty":
        return {"status": "empty", "message": result["message"]}

    result["top_choice"] = _trim(top_choice=result["top_choice"])
    return WSSResult(**{k: result[k] for k in WSSResult.model_fields})


def _trim(*, top_choice: dict) -> dict:
    """Keep only useful identification + score fields for the top choice."""
    keys = [
        "rank", "topsis_score", "s_plus", "s_minus",
        "brand", "model", "variant", "version", "vehicle_type", "powertrain",
        "bike_name", "price_vnd", "curb_weight_kg",
    ]
    return {k: top_choice.get(k) for k in keys if k in top_choice}

File: dss-web/types/dss.ts
/**
 * Kiểu dữ liệu domain của DSS, khớp với response thực tế của backend
 * (dss-service/backend/schemas/dss_schema.py + services/dss_service.py).
 *
 * Orval sinh type từ OpenAPI vào lib/api/generated/model, nhưng FastAPI khai
 * báo `rankings: List[Dict[str, Any]]` nên type sinh ra là `object`. Các kiểu
 * dưới đây bổ sung phần chi tiết mà OpenAPI không mô tả được.
 */

/** 4 loại xe có trong dataset (cột `vehicle_type`). */
export type VehicleType = 'xe tay ga' | 'xe số' | 'xe côn tay' | 'xe máy';

/** Hệ truyền động: ICE = xe xăng, EV = xe điện. */
export type Powertrain = 'ICE' | 'EV' | 'ALL';

/** Khoá của 7 tiêu chí TOPSIS, đúng thứ tự backend mong đợi. */
export type CriterionKey =
  | 'price_vnd'
  | 'fuel_consumption_l_per_100km'
  | 'max_power_kw'
  | 'underseat_storage_l'
  | 'abs'
  | 'curb_weight_kg'
  | 'vehicle_warranty_months';

/** Một tiêu chí trong cấu hình TOPSIS (từ GET /dss/default-criteria). */
export interface CriterionMeta {
  key: CriterionKey;
  name: string;
  type: 'Benefit' | 'Cost';
  benefit: boolean;
  default_weight: number;
  description: string;
}

/** Phần tử của mảng `criteria` trong WSSResult. */
export interface CriterionConfig {
  col: CriterionKey;
  name: string;
  benefit: boolean;
}

/** Body của POST /api/v1/dss/run. */
export interface DSSRunRequest {
  weights?: number[] | null;
  pairwise_matrix?: number[][] | null;
  max_price_vnd?: number | null;
  powertrain?: Powertrain | null;
  brand_list?: string[] | null;
  vehicle_type?: string | null;
}

/**
 * Một dòng trong `rankings`. Backend trả về toàn bộ cột gốc của dataset kèm
 * điểm số, nên mọi trường đều optional trừ rank/score.
 */
export interface RankedBike {
  rank: number;
  topsis_score: number;
  s_plus: number;
  s_minus: number;
  bike_name?: string;
  brand?: string;
  model?: string;
  variant?: string;
  version?: string;
  vehicle_type?: string;
  powertrain?: string;
  price_vnd?: number;
  fuel_consumption_l_per_100km?: number;
  max_power_kw?: number;
  underseat_storage_l?: number;
  abs?: number;
  curb_weight_kg?: number;
  vehicle_warranty_months?: number;
  [key: string]: unknown;
}

/** Thông tin AHP (chỉ có khi gửi pairwise_matrix). */
export interface AHPInfo {
  weights: number[];
  lambda_max: number;
  ci: number;
  cr: number;
  is_consistent: boolean;
}

/** Response thành công của POST /api/v1/dss/run. */
export interface DSSRunSuccess {
  status: 'success';
  total_candidates: number;
  criteria: CriterionConfig[];
  weights: number[];
  rankings: RankedBike[];
  top_choice: Partial<RankedBike>;
  explanation: string;
  ahp?: AHPInfo | null;
  n_criteria?: number;
}

/** Response khi bộ lọc không còn xe nào. */
export interface DSSRunEmpty {
  status: 'empty';
  message: string;
}

export type DSSRunResponse = DSSRunSuccess | DSSRunEmpty;

export function isEmptyResult(res: DSSRunResponse): res is DSSRunEmpty {
  return res.status === 'empty';
}

/** GET /api/v1/dss/default-criteria */
export interface DefaultCriteriaResponse {
  criteria: CriterionMeta[];
  n_criteria: number;
  weights: number[];
  benefit_mask: boolean[];
  source: string;
}

/** GET /api/v1/dss/brands */
export interface BrandsResponse {
  brands: string[];
}

/** GET / */
export interface HealthResponse {
  status: string;
  rows: number;
}
